File: src/main/java/moe/giga/discord/contexts/UserContext.java
package moe.giga.discord.contexts;

import moe.giga.discord.permissions.AccessLevel;
import net.dv8tion.jda.core.entities.Member;
import net.dv8tion.jda.core.entities.User;

public final class UserContext {
    private User user;
    private AccessLevel permissions;

    private ServerContext serverContext;
    private Member member;


    public UserContext(User user) {
        this.user = user;
        this.permissions = AccessLevel.USER;

        this.serverContext = null;
        this.member = null;
    }

    private void resolvePermissions() {
        if (this.serverContext != null && (this.user == this.serverContext.getGuild().getOwner().getUser())) {
            this.permissions = AccessLevel.ADMIN;
        }
    }

    public Member getMember() {
        return member;
    }

    public ServerContext getServerContext() {
        return serverContext;
    }

    public AccessLevel getPermissions() {
        return permissions;
    }

    public User getUser() {
        return user;
    }

    public String getHumanRole() {
        return this.permissions.toString();
    }
}

File: src/main/kotlin/moe/giga/discord/contexts/UserContext.kt
package moe.giga.discord.contexts

import moe.giga.discord.SettingsManager
import moe.giga.discord.util.AccessLevel
import moe.giga.discord.util.username
import net.dv8tion.jda.core.entities.Member
import net.dv8tion.jda.core.entities.User

class UserContext(val user: User, serverContext: ServerContext?) {

    private val permissions: AccessLevel = when {
        user.id == SettingsManager.instance.settings.ownerId -> AccessLevel.ROOT
        serverContext == null -> AccessLevel.USER
        user === serverContext.guild.owner?.user -> AccessLevel.ADMIN
        else -> serverContext.resolvePermissions(user)
    }

    val member: Member? by lazy { serverContext?.getMember(user) }

    val humanRole: String by lazy { this.permissions.toString() }

    val asText: String by lazy { "**${user.username()}**" }

    fun allowed(perms: AccessLevel): Boolean = permissions >= perms
}
